templates/migration-wrapper.sql

```diff
@@ -3,14 +3,19 @@
 DO
 $outer_migration_block$
 DECLARE
-    _already_applied bool = false;
-    _migration       _ps_catalog.migration = row ('{{filename}}', '{{version}}');
+    _migration_name   TEXT = false;
+    _checksum_differs BOOL = false;
+    _migration        _ps_catalog.migration = row ('{{filename}}', '{{version}}', {{checksum}});
 BEGIN
-    SELECT count(*) FILTER (WHERE name = _migration.name) > 0
-    INTO STRICT _already_applied
-    FROM _ps_catalog.migration;
-    IF _already_applied THEN
+    SELECT migration.name, migration.checksum <> _migration.checksum
+    INTO _migration_name, _checksum_differs
+    FROM _ps_catalog.migration
+    WHERE name = _migration.name;
+    IF _migration_name IS NOT NULL THEN
         RAISE LOG 'Migration "{{filename}}" already applied, skipping';
+        IF _checksum_differs THEN
+            RAISE WARNING 'Checksum of migration "{{filename}}" has changed';
```

Comment:
Open question: should we just warn, or fail when discovering a checksum difference?

For our development it would make sense to fail hard here, because changing migration file is in general undesired. But if we would hard fail here then we also need to have a mechanism to purposefully override/rewrite the migration body.

Reply:
What do we tell customers to do in the case it does error out?

Reply:
hmmm... good point. But how do we prevent ourselves from accidentally updating old migrations?

Reply:
Thinking about it more, I tend to agree with you @JamesGuthrie , but I think we need a plan for the case where it happens to a customer

Reply:
@cevian WDYT?

Reply:
I think a warning is good enough here. There /will be/ cases when we need to change migration scripts for future migrations. Our upgrade tests will catch any errors here but let's just warn for now.